webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+type EmbedFieldDict struct {
+	Name   string `json:"name"`
+	Value  string `json:"value"`
+	Inline *bool  `json:"inline,omitempty"`
+}
+
+type DiscordEmbedDict struct {
+	Title       *string             `json:"title,omitempty"`
+	Description *string             `json:"description,omitempty"`
+	URL         *string             `json:"url,omitempty"`
+	Timestamp   *string             `json:"timestamp,omitempty"` // ISO 8601 string
+	Color       *int                `json:"color,omitempty"`
+	Footer      *EmbedFooterDict    `json:"footer,omitempty"`
+	Image       *EmbedImageDict     `json:"image,omitempty"`
+	Thumbnail   *EmbedThumbnailDict `json:"thumbnail,omitempty"`
+	Video       *EmbedVideoDict     `json:"video,omitempty"`
+	Provider    *EmbedProviderDict  `json:"provider,omitempty"`
+	Author      *EmbedAuthorDict    `json:"author,omitempty"`
+	Fields      []EmbedFieldDict    `json:"fields,omitempty"`
+}
+
+// Message represents the incoming message structure from Redis
+type Message struct {
+	ServerNumber  int                `json:"server_number"`
+	Discardable   bool               `json:"discardable"`
+	Edit          bool               `json:"edit"`
+	SentAt        time.Time          `json:"sent_at"`
+	RetryAttempts int                `json:"retry_attempts"`
+	PayloadType   *string            `json:"payload_type,omitempty"`
+	WebhookType   WebhookType        `json:"webhook_type"`
+	MessageType   WebhookMessageType `json:"message_type"`
+	Payload       DiscordWebhookDict `json:"payload"`
+	MessageNumber int
+}
+
+func (m *Message) String() string {
+	return fmt.Sprintf("S#: %d Discard: %t Edit: %t Msg#: %d",
+		m.ServerNumber, m.Discardable, m.Edit, m.MessageNumber)
+}
+
+// Tracks bucket specific rate limits
+type RateLimitState struct {
+	BucketID    string
+	Remaining   int
+	Limit       int
+	ResetTime   time.Time
+	ResetAfter  time.Duration
+	RateLimited bool
+	mu          sync.Mutex
+}
+
+func (r *RateLimitState) String() string {
+	return fmt.Sprintf("ID: %s Remaining: %d Limit: %d ResetTime: %s ResetAfter: %d RateLimited: %t",
+		r.BucketID, r.Remaining, r.Limit, r.ResetTime, r.ResetAfter, r.RateLimited)
+}
+
+// Tracks our local rate limit window
+var localRateLimitState struct {
+	Requests int
+	Window   time.Time
+	mu       sync.Mutex
+}
+
+// Manages a rate limit bucket
+type BucketWorker struct {
+	QueueKey  string
+	rdb       *redis.Client
+	ctx       context.Context
+	rateLimit *RateLimitState
+}
+
+func NewBucketWorker(bucketID string, rdb *redis.Client) *BucketWorker {
+	return &BucketWorker{
+		QueueKey:  bucketQueuePrefix + bucketID,
+		rdb:       rdb,
+		ctx:       context.Background(),
+		rateLimit: &RateLimitState{BucketID: bucketID},
+	}
+}
+
+func GetWorker(bucket string, mu *sync.Mutex) (*BucketWorker, bool) {
+	mu.Lock()
+	defer mu.Unlock()
+	worker, exists := workerLookup[bucket]
+	return worker, exists
+}
+
+func SetWorker(bucket string, worker *BucketWorker, mu *sync.Mutex) {
```

Comment:
If you make `workerLookup` a struct, where the lookup map is a field and the mutex as well, then callers do not need to pass the mutex in.